### PR TITLE
[INLONG-4167][Sort] Add MongoDB extract node

### LINKED ISSUE #4167: [Feature][Sort] Add MongoDB extract node
### Description

[Feature] add Mongo db extract node 

### Use case

[Feature] add Mongo db extract node 

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by EMsnap**
### Title Name: [INLONG-4167][Sort] Add MongoDB extract node

Fixes #4167 

### Motivation

 Add MongoDB extract node

### Modifications

 Add MongoDB extract node

### CODE REVIEW

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    private String collection;
+
+    @JsonCreator
+    public MongoExtractNode(@JsonProperty("id") String id,
+        @JsonProperty("name") String name,
+        @JsonProperty("fields") List<FieldInfo> fields,
+        @Nullable @JsonProperty("watermarkField") WatermarkField waterMarkField,
+        @JsonProperty("properties") Map<String, String> properties,
+        @JsonProperty("primaryKey") String primaryKey,
+        @JsonProperty("tableNames") @Nonnull String collection,
```
Comment (gong) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java:63]: `@JsonProperty("collection")` and `@JsonProperty("tableNames")` is different. So serialization and deserialization will have some problem. Add a UT for serialization and deserialization test in inlong-sort/sort-common/src/test/java/org/apache/inlong/sort/protocol/node/extract

Comment (EMsnap) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java:63]: will do, thanks

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import javax.annotation.Nullable;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonInclude;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonInclude.Include;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.inlong.sort.protocol.FieldInfo;
+import org.apache.inlong.sort.protocol.node.ExtractNode;
+import org.apache.inlong.sort.protocol.transformation.WatermarkField;
+
+public class MongoExtractNode extends ExtractNode implements Serializable {
```
Comment (gong) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java:38]: missing some annotation. 
@EqualsAndHashCode(callSuper = true)
@JsonTypeName("mongoExtract")
@Data
`Node` class need  @JsonSubTypes.Type(value = MongoExtractNode.class, name = "mongoExtract")

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonInclude.Include;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.inlong.sort.protocol.FieldInfo;
+import org.apache.inlong.sort.protocol.node.ExtractNode;
+import org.apache.inlong.sort.protocol.transformation.WatermarkField;
+
+/**
+ * Extract node for mongo, note that mongo should work in replicaSet mode
+ */
+public class MongoExtractNode extends ExtractNode implements Serializable {
```
Comment (yunqingmoswu) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java:38]: 1.Please add the jackson serialization declaration here, and also register the serialization declaration in ExtractNode and Node.
2.It is recommended to add  Data and EqualsAndHashCode annotations, to facilitate obtaining and setting related attribute values.

File: inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/MongoExtractFlinkSqlParseTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.inlong.sort.protocol.node.extract.MongoExtractNode;
+import org.apache.inlong.sort.protocol.node.format.CsvFormat;
+import org.apache.inlong.sort.protocol.node.load.KafkaLoadNode;
+import org.apache.inlong.sort.protocol.transformation.FieldRelationShip;
+import org.apache.inlong.sort.protocol.transformation.relation.NodeRelationShip;
+import org.apache.inlong.sort.singletenant.flink.parser.impl.FlinkSqlParser;
+import org.apache.inlong.sort.singletenant.flink.parser.result.FlinkSqlParseResult;
+import org.junit.Test;
+
+public class MongoExtractFlinkSqlParseTest extends AbstractTestBase {
```
Comment (yunqingmoswu) [inline inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/MongoExtractFlinkSqlParseTest.java:43]: Please add some java comments.

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java (hunk 2 above)
Comment (EMsnap) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/MongoExtractNode.java:38]: done

File: inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/MongoExtractFlinkSqlParseTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env, settings);
+        Node inputNode = buildMongoNode();
+        Node outputNode = buildAllMigrateKafkaNode();
+        StreamInfo streamInfo = new StreamInfo("1", Arrays.asList(inputNode, outputNode),
+            Collections.singletonList(buildNodeRelation(Collections.singletonList(inputNode),
+                Collections.singletonList(outputNode))));
+        GroupInfo groupInfo = new GroupInfo("1", Collections.singletonList(streamInfo));
+        FlinkSqlParser parser = FlinkSqlParser.getInstance(tableEnv, groupInfo);
+        FlinkSqlParseResult result = parser.parse();
+        result.execute();
```
[the remark this answers is not in the document]
Comment (EMsnap) [reply inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/MongoExtractFlinkSqlParseTest.java:101]: done

**Review by yunqingmoswu — APPROVED** [review]

**Review by yunqingmoswu — APPROVED** [review]

**Review by GanfengTan — APPROVED** [review]

**Review by gong — APPROVED** [review]

**Review by healchow — APPROVED** [review]

**Pull request merged**